### ISSUE TITLE
Rule suggestion: Don't reassign function parameters

### ISSUE THREAD

**Issue by lxanders** [eslint/eslint#1599]
Reusing function parameters feels often like a code smell for me. If you consider something like the following example it could even cost you some debugging time (especially if you have long legacy functions):

```js
function foo (bar) {
    bar = 13;
    // now we accidentally overwrote the function parameter which is most likely not good
}
```

I suggest the implementation of a rule that is similar to [no-ex-assign](http://eslint.org/docs/rules/no-ex-assign.html) but for functions instead of try-catch-blocks and is turned off by default.

What do you think about it?

**Comment by bedney** [comment]
Completely agree on all counts. They should be separate rules and they are both code smells, IMHO.

**Comment by burnnat** [comment]
I'm working on this, should have a PR in the next day or two.

**Comment by nzakas** [comment]
Awesome!

**Comment by xjamundx** [comment]
What about things like:

```js
function(boo) {

    // basic variations of default parameters
    boo = boo || 13;
    boo = boo === undefined ? 13 : boo;
    boo = typeof boo === 'undefined' ? 13 : boo;

    // defaults for an options object
    Object.assign(boo, {
      value: 13
    });
}
```

**Comment by lo1tuma** [comment]
@xjamundx I would consider your examples as a code smell and would prefer the rule to warn on such patterns. If you wan’t to reassign function parameters, you can simple disable this rule.

**Comment by xjamundx** [comment]
I'd just suggest if that's intentional to specifically test and document that the default parameters style is considered bad. I think for many of us, it might be a surprise, though not necessarily worth carving out an exception for.

In ES6 environments the answer will just be put it in the functional declaration anyway.

**Comment by bedney** [comment]
@lo1tuma - Yep. The exact purpose of this rule is to point out that this is a code smell.

**Comment by nzakas** [comment]
@xjamundx many do consider that as bad. This is why rules are configurable, if you disagree, you can turn it off. :)

**Comment by idchlife** [comment]
Guys, but what about .map? Lint saying this is bad, when I use reassignment in .map

**Comment by ilyavolodin** [comment]
Why would you ever want to do reassignment in `map`? The whole point of `map` is to translate from one array to another, not to mutate the original array.

**Comment by idchlife** [comment]
@ilyavolodin 

You mean when I'm doing like this:

    surveyWithIds.questions = survey.questions.map((q: QuestionType): QuestionType => {
      q.id = uniqid();

      if (q.type === 'choice_list') {
        q.choices = q.choices.map((c: ChoiceType): ChoiceType => {
          c.id = uniqid();

          return c;
        });
      }

      return q;
    });

It is considered bad practice? Old surveyWithIds.questions is gone anyway :|

**Comment by xjamundx** [comment]
Yeah so for example, what about `q.choices.map(c => { ...c, id: uniqid() })`

**Comment by idchlife** [comment]
@xjamundx I suggest you to fix your example by adding additional curly brackets and return statement, because show function syntax won't work that way.

Yep, that's actually the stuff. Thanks!

**Comment by xjamundx** [comment]
@idchlife this is valid in chrome `[1,2].map(c => ({a: 1}))` not sure about TypeScript?

**Comment by idchlife** [comment]
@xjamundx oh, I see. I did not know about () around {} when returning object in short function.

That's actually just flowtype, not typescript.

**Comment by PaulDejardin** [comment]
I agree with the previous comments here that:

- reassigning arguments is a code smell
- reassigning named arguments is a code smell
- modifying properties (and nested properties) of named arguments is a code smell.

I'm currently evaluating no-param-reassign for a fairly large code base, because it's easy to perpetuate a pattern of mutating a single large argument blob over time, but it makes the code quite difficult to read and debug. The rule matches over 400 times, 90% of which could be considered extraneous to the core logic (stuff like assigning things to the global object, or to a parameter called scope (this is an Angular app). In practice that means it is difficult to employ this rule because those few arguments are "OK" to modify for structural reasons which would be difficult to refactor.

I'd like to propose adding another parameter to this rule for excluding specific property paths.

```
{
  "error",
  {
    "props": true,
    "exclude": {
      "global": false,
      "global.library": true,
      "scope": true,
      "some.other.specific.path.two": false
    }
  }
}
```

**Comment by platinumazure** [comment]
Can we please create a new issue for enhancement requests to this rule? Thanks!
